state/executor.go

```diff
@@ -27,6 +28,31 @@ type GetHashByNumber = func(i uint64) types.Hash
 
 type GetHashByNumberHelper = func(*types.Header) GetHashByNumber
 
+type StakingEventType string
+
+const (
+	StakingEventStaked   StakingEventType = "Staked"
+	StakingEventUnstaked StakingEventType = "Unstaked"
+)
+
+type StakingEvent struct {
+	Number  int64
+	Type    StakingEventType
+	Address types.Address
+	Amount  *big.Int
+}
+
+type StakingEventSubscription struct {
+	EventCh chan *StakingEvent
```

Comment:
This can be one-directional channel if I'm not mistaken:
```suggestion
	EventCh chan<- *StakingEvent
```

Reply:
Can't change the type to one-directional channel here. Instead of this, made this field private and add method to get channel to receive event. Other package won't be able to send events into this channel. Accordingly, I change `WaitForDone` method to `Close`. It close own channel after all transmissions finishes.

Reply:
Oh, I see. You didn't have to do that, I thought that it could be made one-directional, but as it seems I was mistaken. 🙂